swiftide/src/transformers/contextualize_code_chunk.rs

```diff
@@ -0,0 +1,181 @@
+    prompt_template: PromptTemplate,
+    #[builder(default)]
+    concurrency: Option<usize>,
+}
+
+fn extract_markdown_codeblock(text: String) -> String {
+    let re = regex::Regex::new(r"(?sm)```\w*\n(.*?)```").unwrap();
+    let captures = re.captures(text.as_str());
+    captures
+        .map(|c| c.get(1).unwrap().as_str().to_string())
+        .unwrap_or(text)
+}
+
+impl ContextualizeCodeChunk {
+    pub fn builder() -> ContextualizeCodeChunkBuilder {
+        ContextualizeCodeChunkBuilder::default()
+    }
+
+    pub fn from_client(client: impl SimplePrompt + 'static) -> ContextualizeCodeChunkBuilder {
+        ContextualizeCodeChunkBuilder::default()
+            .client(client)
+            .to_owned()
+    }
+    /// Creates a new instance of `ContextualizeCodeChunk`.
+    ///
+    /// # Arguments
+    ///
+    /// * `client` - An implementation of the `SimplePrompt` trait used to generate questions and answers.
+    ///
+    /// # Returns
+    ///
+    /// A new instance of `ContextualizeCodeChunk` with a default prompt and a default number of questions.
+    pub fn new(client: impl SimplePrompt + 'static) -> Self {
+        Self {
+            client: Arc::new(client),
+            prompt_template: default_prompt(),
+            concurrency: None,
+        }
+    }
+
+    #[must_use]
+    pub fn with_concurrency(mut self, concurrency: usize) -> Self {
+        self.concurrency = Some(concurrency);
+        self
+    }
+}
+
+/// Returns the default prompt template for generating questions and answers.
+///
+/// This template includes placeholders for the number of questions and the code chunk.
+///
+/// # Returns
+///
+/// A string representing the default prompt template.
+fn default_prompt() -> PromptTemplate {
+    PromptTemplate::from_compiled_template_name("contextualize_code_chunk.prompt.md")
+}
+
+impl ContextualizeCodeChunkBuilder {
+    pub fn client(&mut self, client: impl SimplePrompt + 'static) -> &mut Self {
+        self.client = Some(Arc::new(client));
+        self
+    }
+}
+
+#[async_trait]
+impl Transformer for ContextualizeCodeChunk {
+    /// Asynchronously transforms an `Node` by adding context to it based on the original file metadata.
+    ///
+    /// This method uses the `SimplePrompt` client to merge the chunk with its context.
+    ///
+    /// # Arguments
+    ///
+    /// * `node` - The `Node` to be transformed.
+    ///
+    /// # Returns
+    ///
+    /// A result containing the transformed `Node` or an error if the transformation fails.
+    ///
+    /// # Errors
+    ///
+    /// This function will return an error if the `SimplePrompt` client fails to generate a response.
+    #[tracing::instrument(skip_all, name = "transformers.contextualize_code_chunk")]
+    async fn transform_node(&self, mut node: Node) -> Result<Node> {
+        let maybe_original_size = node
+            .metadata
+            .get("Original Size")
+            .map(|size| size.parse::<usize>().unwrap());
```

Comment:
Proper values would also remove the parse and the unwrap

Reply:
Yeah that would be nice. If you think having these values on all nodes makes sense then that would definitely make it more clean.

Reply:
I think the size of the original chunk and a chunks offset in the chunk could be useful values after any chunking.